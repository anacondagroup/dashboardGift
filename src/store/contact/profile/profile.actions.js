import { PROFILE_LOAD_REQUEST, PROFILE_LOAD_SUCCESS, PROFILE_LOAD_FAIL, RESET_PROFILE_STATE } from './profile.types';
import { transformContactInfo } from './profile.helpers';

export const profileLoadRequest = ({ giftId = null, contactId = null }) => ({
  type: PROFILE_LOAD_REQUEST,
  payload: { giftId, contactId },
});

export const profileLoadSuccess = profile => ({
  type: PROFILE_LOAD_SUCCESS,
  payload: transformContactInfo(profile),
});

export const profileLoadFail = error => ({
  type: PROFILE_LOAD_FAIL,
  payload: error,
});

export const resetProfile = () => ({
  type: RESET_PROFILE_STATE,
});
