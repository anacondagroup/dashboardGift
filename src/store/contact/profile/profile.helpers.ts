/* eslint-disable camelcase */
import { applySpec, pathOr, prop, path, pipe, unless, isNil, converge, or } from 'ramda';

export type TContactResearchInfo = {
  image: string;
  company: string;
  position: string;
  education: string;
  location: string;
  discoveryNotes: string;
  researchTags: { title: string; url: string; tags: string[] }[];
};

export type TContactInfo = {
  email: string;
  firstName: string;
  lastName: string;
  fullName: string;
  countryId: number;
  isUnsubscribe: boolean;
  company: string;
  research: TContactResearchInfo | null;
};

export type TContactStatistic = {
  created: number;
  sent: number;
  accepted: number;
  meetings: number;
};

export type TContactLatestGift = {
  id: number;
  status_id: number;
  history: { label: string; status_string: string; action_date: string }[];
  shipping_info: {
    tracking_carrier: string;
    tracking_number: string;
    tracking_state: string;
    tracking_link: string;
  };
  need_more_info_message: string;
  prospect_meeting_notes: string;
  free_meetings: { start: string; end: string; timezone: string }[];
  sent_by: {
    id: number;
    full_name: string;
    avatar: string;
    email: string;
    has_integration: boolean;
  };
  sent_as: {
    id: number;
    full_name: string;
    avatar: string;
    email: string;
    has_integration: boolean;
  };
  campaign: {
    id: number;
    name: string;
    team_id: number;
  };
  product: {
    name: string;
    image: string;
  };
  meeting: {
    id: number;
    start: string;
    end: string;
    title: string;
    meeting_type: string;
    additional_info: string;
    outcome: string;
    status: string;
  };
};

export type TContact = {
  contact: TContactInfo;
  statistic: TContactStatistic;
  latestGift: TContactLatestGift | null;
};

const transformContactGift = unless(
  isNil,
  applySpec<TContactLatestGift>({
    sentAs: prop('sent_as'),
    sentBy: prop('sent_by'),
    shippingInfo: prop('shipping_info'),
    statusId: prop('status_id'),
    adminNote: prop('admin_note'),
    id: prop('id'),
    history: prop('history'),
    need_more_info_message: prop('need_more_info_message'),
    prospect_meeting_notes: prop('prospect_meeting_notes'),
    free_meetings: prop('free_meetings'),
    campaign: prop('campaign'),
    product: prop('product'),
    meeting: prop('meeting'),
  }),
);

// TODO: Migrate profile store to TS
export const transformContactInfo = applySpec({
  id: path(['contact', 'id']),
  email: path(['contact', 'email']),
  avatar: pathOr('', ['contact', 'research', 'image']),
  company: converge(or, [path(['contact', 'research', 'company']), path(['contact', 'company'])]),
  employment: converge(or, [path(['contact', 'position']), path(['contact', 'research', 'employment'])]),
  firstName: path(['contact', 'firstName']),
  lastName: path(['contact', 'lastName']),
  fullName: path(['contact', 'fullName']),
  education: pathOr('', ['contact', 'research', 'education']),
  location: pathOr('', ['contact', 'research', 'location']),
  phoneNumber: path(['contact', 'phone']),
  discoveryNotes: pathOr('', ['contact', 'research', 'discoveryNotes']),
  researchTags: pathOr([], ['contact', 'research', 'researchTags']),
  countryId: path(['contact', 'countryId']),
  statistic: prop('statistic'),
  latestGift: pipe(prop('latestGift'), transformContactGift),
  currentGift: pipe(prop('currentGift'), transformContactGift),
});
