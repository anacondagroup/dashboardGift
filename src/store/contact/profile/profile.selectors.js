import * as R from 'ramda';

import { getContactState } from '../contact.selectors';
import { viewLoading } from '../../../helpers/lens.helpers';

export const getProfileState = R.compose(R.prop('profile'), getContactState);

export const getProfile = getProfileState;

export const getStatistic = R.compose(R.pick(['statistic']), getProfileState);

export const getLatestGift = R.compose(R.pick(['latest_gift']), getProfileState);

export const getProfileIsLoading = R.compose(viewLoading, getProfileState);
