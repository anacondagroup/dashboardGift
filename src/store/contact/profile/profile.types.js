export const PROFILE_LOAD_REQUEST = 'PROFILE_LOAD_REQUEST';
export const PROFILE_LOAD_SUCCESS = 'PROFILE_LOAD_SUCCESS';
export const PROFILE_LOAD_FAIL = 'PROFILE_LOAD_FAIL';
export const RESET_PROFILE_STATE = 'RESET_PROFILE_STATE';
