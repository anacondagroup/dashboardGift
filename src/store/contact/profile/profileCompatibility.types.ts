export interface IResearchTag {
  title: string;
  url: string;
  tags: string[];
}

export interface IProfile {
  id: number | null;
  avatar: string;
  email: string;
  firstName: string;
  fullName: string;
  lastName: string;
  employment: string;
  education: string;
  location: string;
  phoneNumber: string;
  discoveryNotes?: string;
  researchTags: IResearchTag[];
  countryId?: number;
  isUnsubscribed: boolean;
}

/* eslint-disable camelcase */
export interface IGiftRecipient {
  id: number | null;
  first_name: string;
  last_name: string;
  email: string;
  full_name: string;
  employment: string;
  company: string;
  avatar: string;
}
/* eslint-enable camelcase */
