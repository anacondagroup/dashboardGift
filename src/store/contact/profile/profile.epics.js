import { ofType } from 'redux-observable';
import { catchError, map, mergeMap, debounce } from 'rxjs/operators/index';
import { of, timer } from 'rxjs';
import { ContactDetails, updateSearch } from '@alycecom/modules';
import { push } from 'connected-react-router';

import { PROFILE_LOAD_REQUEST } from './profile.types';
import { profileLoadSuccess, profileLoadFail } from './profile.actions';

export const loadProfileEpic = (action$, state$, { apiService }) =>
  action$.pipe(
    ofType(PROFILE_LOAD_REQUEST),
    debounce(() => timer(300)),
    mergeMap(({ payload: { contactId, giftId } }) =>
      (giftId
        ? apiService.get(`/enterprise/gifts/${giftId}/contact-info`, {}, true)
        : apiService.get(`/enterprise/contact/${contactId}/info`, {}, true)
      ).pipe(
        map(response => profileLoadSuccess(response.data)),
        catchError(error => of(profileLoadFail(error))),
      ),
    ),
  );

export const needMoreInfoSuccessEpic = (action$, state$) =>
  action$.pipe(
    ofType(ContactDetails.actionTypes.SEND_MORE_INFO_SUCCESS),
    mergeMap(({ payload }) => [
      push({
        pathname: state$.value.router.location.pathname,
        search: updateSearch(state$.value.router.location.search, { sidebar_tab: 'send-gift', gift_id: payload }),
      }),
    ]),
  );

export const profileEpics = [loadProfileEpic, needMoreInfoSuccessEpic];
