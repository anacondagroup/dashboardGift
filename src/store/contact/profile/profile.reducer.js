import { PROFILE_LOAD_REQUEST, PROFILE_LOAD_SUCCESS, PROFILE_LOAD_FAIL, RESET_PROFILE_STATE } from './profile.types';

export const initialState = {
  isLoading: false,
  error: null,
  id: null,
  avatar: '',
  email: '',
  firstName: '',
  fullName: '',
  lastName: '',
  employment: '',
  education: '',
  location: '',
  phoneNumber: '',
  discoveryNotes: null,
  researchTags: [],
  countryId: null,
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case PROFILE_LOAD_REQUEST:
      return {
        ...state,
        isLoading: true,
      };
    case PROFILE_LOAD_SUCCESS:
      return {
        ...state,
        error: null,
        isLoading: false,
        ...action.payload,
      };
    case PROFILE_LOAD_FAIL:
      return {
        ...state,
        isLoading: false,
        error: action.payload,
      };
    case RESET_PROFILE_STATE:
      return initialState;
    default:
      return state;
  }
};
