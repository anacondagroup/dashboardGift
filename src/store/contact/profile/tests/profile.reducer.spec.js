import { reducer, initialState } from '../profile.reducer';
import { profileLoadSuccess } from '../profile.actions';

describe('Profile reducer', () => {
  describe('load success', () => {
    it('Should rename keys', () => {
      const actual = reducer(
        initialState,
        profileLoadSuccess({
          contact: {
            id: 1429,
            email: 'barr_tim@grandhotels.com',
            firstName: 'Bler',
            lastName: 'Barr',
            fullName: 'Bler Barr',
            isUnsubscribed: false,
            phone: '',
            discoveryNotes: '',
            countryId: 1,
            position: 'SVP, Administration and Finance, Grand Hotels & Resorts Ltd',
            research: {
              company: 'Cmp Ltd.',
              image: 'http://alyce.site/images/avatar.png',
              education: '',
              location: 'US asa 123 US',
              researchTags: [],
            },
          },
          latestGift: null,
          currentGift: null,
          statistic: {
            created: 0,
            sent: 0,
            accepted: 0,
            meetings: 0,
          },
        }),
      );

      expect(actual).toMatchSnapshot();
    });
  });
});
