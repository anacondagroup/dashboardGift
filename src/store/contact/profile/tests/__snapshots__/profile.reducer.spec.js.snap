// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile reducer load success Should rename keys 1`] = `
Object {
  "avatar": "http://alyce.site/images/avatar.png",
  "company": "Cmp Ltd.",
  "countryId": 1,
  "currentGift": null,
  "discoveryNotes": "",
  "education": "",
  "email": "barr_tim@grandhotels.com",
  "employment": "SVP, Administration and Finance, Grand Hotels & Resorts Ltd",
  "error": null,
  "firstName": "Bler",
  "fullName": "Bler Barr",
  "id": 1429,
  "isLoading": false,
  "lastName": "Barr",
  "latestGift": null,
  "location": "US asa 123 US",
  "phoneNumber": "",
  "researchTags": Array [],
  "statistic": Object {
    "accepted": 0,
    "created": 0,
    "meetings": 0,
    "sent": 0,
  },
}
`;
